### ISSUE TITLE
Bug report: "ConstructorInvokesOverridable" liberally errors on `@Override` method in enum called by enum constructor

### ISSUE THREAD

**Issue by jbduncan** [google/error-prone#709]
# Overview

As I was working on introducing error-prone to JUnit 5, I came across the following strange behaviour, which I believe is a bug in the ["ConstructorInvokesOverridable"](http://errorprone.info/bugpattern/ConstructorInvokesOverridable) bug pattern.

If given the following source file (taken and shortened from https://github.com/jbduncan/junit5/blob/error-prone-investigation-3/junit-platform-console/src/main/java/org/junit/platform/console/options/Theme.java), error-prone considers the constructor to be calling an overridable method and thus reports an error, stopping all compilation. Usually this would be a valid security-related bug in the code, but in this case, the class is an enum, which AFAIK cannot be subclassed, thus meaning the offending method `vertical()` cannot be overridden.

We're working around this in the meantime by making `Theme.vertical()` final.

```java
public enum Theme {

	ASCII(".", "| ", "+--", "'--", "[OK]", "[A]", "[X]", "[S]"),
	UNICODE("╷", "│  ", "├─", "└─", "✔", "■", "✘", "↷");

	public static Theme valueOf(Charset charset) {
		if (StandardCharsets.UTF_8.equals(charset)) {
			return UNICODE;
		}
		return ASCII;
	}

	private final String[] tiles;
	private final String blank;

	Theme(String... tiles) {
		this.tiles = tiles;
		this.blank = new String(new char[vertical().length()]).replace('\0', ' ');
	}

	public String root() {
		return tiles[0];
	}

	public String vertical() {
		return tiles[1];
	}

	public String blank() {
		return blank;
	}

	public String entry() {
		return tiles[2];
	}

	public String end() {
		return tiles[3];
	}

	public String successful() {
		return tiles[4];
	}

	public String aborted() {
		return tiles[5];
	}

	public String failed() {
		return tiles[6];
	}

	public String skipped() {
		return tiles[7];
	}

	public String status(TestExecutionResult result) {
		switch (result.getStatus()) {
			case SUCCESSFUL:
				return successful();
			case ABORTED:
				return aborted();
			case FAILED:
				return failed();
			default:
				return result.getStatus().name();
		}
	}
	@Override
	public String toString() {
		return name().toLowerCase();
	}
}
```

# How to reproduce
1. Clone https://github.com/jbduncan/junit5 locally.
2. Checkout branch "error-prone-investigation-3".
3. Run `./gradlew check` from the root.

# Expected output
A successful compilation is expected.

# Actual output
```
C:\Users\Jonathan\dev\Java\IntelliJ Projects\junit5\junit-platform-console\src\main\java\org\junit\platform\console\options\Theme.java:81: warning: [ConstructorInvokesOverridable] Constructors should not invoke overridable methods.
		this.blank = new String(new char[vertical().length()]).replace('\0', ' ');
		                                         ^
    (see http://errorprone.info/bugpattern/ConstructorInvokesOverridable)
```

**Comment by cushon** [comment]
> the class is an enum, which AFAIK cannot be subclassed

What about:

```java
public enum Theme {
  ASCII(".", "| ", "+--", "'--", "[OK]", "[A]", "[X]", "[S]") {
    @Override public String vertical() { ... }
  } ,
  ...
}
```

> reports an error, stopping all compilation

Note that this check is a warning, so it should not stop compilation unless you're using `-Werror` or have explicitly enabled this check as an error.

**Comment by jbduncan** [comment]
> Note that this check is a warning, so it should not stop compilation unless you're using -Werror or have explicitly enabled this check as an error.

Yes, we're using `-Werror` over at JUnit 5 to make production Java source files fail upon producing any sort of warning. I don't know the rationale behind enabling it, so I don't think it's terribly practical for us to disable it for now.

@sbrannen @marcphillip Do you have any insight on why `-Werror` was decided to be enabled for JUnit 5's build?

**Comment by jbduncan** [comment]
> What about:
>
> ```java
> public enum Theme {
>   ASCII(".", "| ", "+--", "'--", "[OK]", "[A]", "[X]", "[S]") {
>     @Override public String vertical() { ... }
>   } ,
>   ...
> }
> ```

Oh, clever! I didn't realise one could override methods in enums like that.

I can confirm that overriding is happening, because when I make `vertical()` final, the `@Override public String vertical() { ... }` within `ASCII` becomes a compilation error.

Looks like I'm mistaken then that methods in enums cannot be overridden!

**Comment by jbduncan** [comment]
I'll close this issue for now.
